flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -135,26 +145,31 @@ private static List<InternalResourceInfo> getPendingResources(
                 .collect(Collectors.toList());
     }
 
-    private static int tryFulfilledRequirementWithResource(
+    private int tryFulfilledRequirementWithResource(
```

Comment:
Maybe we should make the entire `tryFulfilledRequirementWithResource` the strategy method.

For the any matching strategy, we want to allocate as many slots as possible for each selected TM. And for pending resources, we should always use this strategy, so that we use as less pending workers as possible and the rest can be canceled.

For the least utilization strategy, we can use a priority queue to reduce the overhead of rapidly calling `min()` on the entire worker list.